File: scripts/serve_ui.mjs
import { createServer } from 'node:http';
import { promises as fs } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const SCRIPT_DIR = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(SCRIPT_DIR, '..');
const UI_DIR = path.resolve(REPO_ROOT, 'ui');
const ARTIFACTS_DIR = path.resolve(REPO_ROOT, 'artifacts');

function parseArgs(argv) {
  const out = { port: 4173 };
  for (const arg of argv) {
    if (arg.startsWith('--port=')) {
      const n = Number(arg.split('=')[1]);
      if (Number.isFinite(n) && n > 0) out.port = n;
    }
  }
  return out;
}

async function exists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function walkFiles(rootDir) {
  const out = [];
  async function walk(dir) {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.resolve(dir, entry.name);
      if (entry.isDirectory()) await walk(full);
      else out.push(full);
    }
  }
  if (await exists(rootDir)) await walk(rootDir);
  return out;
}

async function latestByName(filename) {
  const files = (await walkFiles(ARTIFACTS_DIR)).filter((p) => path.basename(p) === filename);
  if (!files.length) return null;
  const stats = await Promise.all(files.map(async (p) => ({ p, s: await fs.stat(p) })));
  stats.sort((a, b) => b.s.mtimeMs - a.s.mtimeMs);
  return stats[0].p;
}

async function discoverArtifactPaths() {
  const leaderboardPath =
    (await latestByName('leaderboard.frontend.json')) ||
    (await latestByName('latest.frontend.json')) ||
    (await latestByName('latest.json'));

  const debugPath = (await latestByName('latest.debug.json')) || (await latestByName('snapshot.debug.json'));

  return { leaderboardPath, debugPath };
}

function json(res, status, payload) {
  const body = `${JSON.stringify(payload, null, 2)}\n`;
  res.writeHead(status, { 'content-type': 'application/json; charset=utf-8' });
  res.end(body);
}

function text(res, status, body, contentType = 'text/plain; charset=utf-8') {
  res.writeHead(status, { 'content-type': contentType });
  res.end(body);
}

function contentTypeFor(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === '.html') return 'text/html; charset=utf-8';
  if (ext === '.js') return 'text/javascript; charset=utf-8';
  if (ext === '.css') return 'text/css; charset=utf-8';
  if (ext === '.json') return 'application/json; charset=utf-8';
  if (ext === '.pdf') return 'application/pdf';
  return 'application/octet-stream';
}

async function sendFile(res, filePath) {
  try {
    const data = await fs.readFile(filePath);
    res.writeHead(200, { 'content-type': contentTypeFor(filePath) });
    res.end(data);
  } catch {
    text(res, 404, 'Not found');
  }
}

function safeResolveRepoPath(inputPath) {
  const abs = path.resolve(REPO_ROOT, inputPath);
  if (!abs.startsWith(REPO_ROOT)) return null;
  return abs;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const artifacts = await discoverArtifactPaths();

  const server = createServer(async (req, res) => {
    try {
      const url = new URL(req.url || '/', 'http://localhost');
      const pathname = url.pathname;

      if (pathname === '/api/meta') {
        return json(res, 200, {
          repo_root: REPO_ROOT,
          artifacts_dir: ARTIFACTS_DIR,
          leaderboard_path: artifacts.leaderboardPath,
          debug_path: artifacts.debugPath,
        });
      }

      if (pathname === '/api/leaderboard') {
        if (!artifacts.leaderboardPath) return json(res, 404, { error: 'No leaderboard artifact found.' });
        const raw = await fs.readFile(artifacts.leaderboardPath, 'utf8');
        res.writeHead(200, { 'content-type': 'application/json; charset=utf-8' });
        return res.end(raw);
      }

      if (pathname === '/api/debug') {
        if (!artifacts.debugPath) return json(res, 404, { error: 'No debug artifact found.' });
        const raw = await fs.readFile(artifacts.debugPath, 'utf8');
        res.writeHead(200, { 'content-type': 'application/json; charset=utf-8' });
        return res.end(raw);
      }

      if (pathname === '/api/document') {
        const sourcePdf = url.searchParams.get('source_pdf') || '';
        const target = safeResolveRepoPath(sourcePdf);
        if (!target) return json(res, 400, { error: 'Invalid source_pdf path.' });
        if (!(await exists(target))) return json(res, 404, { error: `Document not found: ${sourcePdf}` });
        return sendFile(res, target);
      }

      if (pathname === '/' || pathname === '/index.html') return sendFile(res, path.resolve(UI_DIR, 'index.html'));
      if (pathname === '/debug' || pathname === '/debug.html') return sendFile(res, path.resolve(UI_DIR, 'debug.html'));

      const uiFile = safeResolveRepoPath(path.relative(REPO_ROOT, path.resolve(UI_DIR, pathname.slice(1))));
      if (uiFile && uiFile.startsWith(UI_DIR) && (await exists(uiFile))) {
        return sendFile(res, uiFile);
      }

      text(res, 404, 'Not found');
    } catch (error) {
      json(res, 500, { error: error instanceof Error ? error.message : 'Internal server error.' });
    }
  });

  server.listen(args.port, '127.0.0.1', () => {
    console.log(`OCR mini-bench UI available at http://127.0.0.1:${args.port}`);
    console.log(`Leaderboard source: ${artifacts.leaderboardPath || 'not found'}`);
    console.log(`Debug source: ${artifacts.debugPath || 'not found'}`);
  });
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/serve_ui.py
"""Tiny HTTP server for the OCR mini-bench UI. Python port of
`scripts/serve_ui.mjs`, using stdlib `http.server` to avoid a FastAPI dep
for a static-file + JSON-endpoint app.

Routes:
    GET /api/meta         - paths the server resolved
    GET /api/leaderboard  - latest leaderboard.frontend.json (or fallbacks)
    GET /api/debug        - latest latest.debug.json (or fallback)
    GET /api/document?source_pdf=PATH - serve a PDF from bench_documents
    GET /                 - ui/index.html
    GET /debug            - ui/debug.html
    GET /<file>           - static files under ui/

Usage:
    uv run python scripts/serve_ui.py [--port=4173]
"""

from __future__ import annotations

import argparse
import contextlib
import json
import sys
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from typing import ClassVar
from urllib.parse import parse_qs, urlparse

from ocr_mini_bench.config.paths import PATHS

REPO_ROOT = PATHS.dataset.manifest.parent.parent
UI_DIR = REPO_ROOT / "ui"
ARTIFACTS_DIR = REPO_ROOT / "artifacts"


def _walk_files(root: Path) -> list[Path]:
    if not root.exists():
        return []
    return [p for p in root.rglob("*") if p.is_file()]


def _latest_by_name(filename: str) -> Path | None:
    matches = [p for p in _walk_files(ARTIFACTS_DIR) if p.name == filename]
    if not matches:
        return None
    matches.sort(key=lambda p: p.stat().st_mtime, reverse=True)
    return matches[0]


def _discover_artifact_paths() -> dict[str, Path | None]:
    leaderboard = (
        _latest_by_name("leaderboard.frontend.json")
        or _latest_by_name("latest.frontend.json")
        or _latest_by_name("latest.json")
    )
    debug = _latest_by_name("latest.debug.json") or _latest_by_name("snapshot.debug.json")
    return {"leaderboardPath": leaderboard, "debugPath": debug}


_CONTENT_TYPES = {
    ".html": "text/html; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".pdf": "application/pdf",
}


def _content_type_for(file_path: Path) -> str:
    return _CONTENT_TYPES.get(file_path.suffix.lower(), "application/octet-stream")


def _safe_resolve_repo_path(input_path: str) -> Path | None:
    """Resolve `input_path` against REPO_ROOT, refusing any result that
    escapes the repo. Path traversal guard."""
    if not input_path:
        return None
    abs_path = (REPO_ROOT / input_path).resolve()
    try:
        abs_path.relative_to(REPO_ROOT.resolve())
    except ValueError:
        return None
    return abs_path


class _Handler(BaseHTTPRequestHandler):
    artifacts: ClassVar[dict[str, Path | None]] = {}

    def log_message(self, format: str, *args: object) -> None:
        # Quieter default than the stdlib's stderr-per-request spam.
        return

    def _send_json(self, status: int, payload: object) -> None:
        body = (json.dumps(payload, indent=2) + "\n").encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _send_text(self, status: int, body: str, content_type: str = "text/plain; charset=utf-8") -> None:
        data = body.encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _send_file(self, file_path: Path) -> None:
        try:
            data = file_path.read_bytes()
        except OSError:
            self._send_text(404, "Not found")
            return
        self.send_response(200)
        self.send_header("Content-Type", _content_type_for(file_path))
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def _send_raw_json(self, file_path: Path) -> None:
        try:
            data = file_path.read_bytes()
        except OSError:
            self._send_json(404, {"error": "Read failed."})
            return
        self.send_response(200)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def do_GET(self) -> None:
        try:
            parsed = urlparse(self.path or "/")
            pathname = parsed.path

            if pathname == "/api/meta":
                self._send_json(
                    200,
                    {
                        "repo_root": str(REPO_ROOT),
                        "artifacts_dir": str(ARTIFACTS_DIR),
                        "leaderboard_path": str(self.artifacts.get("leaderboardPath"))
                        if self.artifacts.get("leaderboardPath")
                        else None,
                        "debug_path": str(self.artifacts.get("debugPath"))
                        if self.artifacts.get("debugPath")
                        else None,
                    },
                )
                return

            if pathname == "/api/leaderboard":
                p = self.artifacts.get("leaderboardPath")
                if not p:
                    self._send_json(404, {"error": "No leaderboard artifact found."})
                    return
                self._send_raw_json(p)
                return

            if pathname == "/api/debug":
                p = self.artifacts.get("debugPath")
                if not p:
                    self._send_json(404, {"error": "No debug artifact found."})
                    return
                self._send_raw_json(p)
                return

            if pathname == "/api/document":
                params = parse_qs(parsed.query)
                source_pdf = params.get("source_pdf", [""])[0]
                target = _safe_resolve_repo_path(source_pdf)
                if not target:
                    self._send_json(400, {"error": "Invalid source_pdf path."})
                    return
                if not target.exists():
                    self._send_json(404, {"error": f"Document not found: {source_pdf}"})
                    return
                self._send_file(target)
                return

            if pathname in ("/", "/index.html"):
                self._send_file(UI_DIR / "index.html")
                return
            if pathname in ("/debug", "/debug.html"):
                self._send_file(UI_DIR / "debug.html")
                return

            ui_candidate = (UI_DIR / pathname.lstrip("/")).resolve()
            try:
                ui_candidate.relative_to(UI_DIR.resolve())
            except ValueError:
                self._send_text(404, "Not found")
                return
            if ui_candidate.exists() and ui_candidate.is_file():
                self._send_file(ui_candidate)
                return

            self._send_text(404, "Not found")
        except Exception as e:
            self._send_json(500, {"error": str(e) or "Internal server error."})


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(add_help=True)
    parser.add_argument("--port", type=int, default=4173)
    args = parser.parse_args(argv)

    artifacts = _discover_artifact_paths()
    _Handler.artifacts = artifacts

    server = HTTPServer(("127.0.0.1", args.port), _Handler)
    print(f"OCR mini-bench UI available at http://127.0.0.1:{args.port}")
    print(f"Leaderboard source: {artifacts['leaderboardPath'] or 'not found'}")
    print(f"Debug source: {artifacts['debugPath'] or 'not found'}")
    with contextlib.suppress(KeyboardInterrupt):
        server.serve_forever()
    return 0


if __name__ == "__main__":
    sys.exit(main())
